File: frontend/src/lib/edgedepth/core/heatmap_manager.ts
// heatmap_manager.h/.cpp — exact port line by line, space by space, bracket by bracket, as is
// Original: https://github.com/edgedepthhq/edgedepth-terminal.git src/core/heatmap_manager.h/.cpp
// Implemented as is into LSE — functional orderflow, GPU ring 8192x1024 LUT 256x1, SoA

import { ShaderHeatmapRenderer } from '../rendering/shader_heatmap_renderer';

export type HeatmapKey = { exchange:string; symbol:string; mode:string; };

export class HeatmapManager {
  private heatmaps = new Map<string, ShaderHeatmapRenderer>();

  private keyToString(k:HeatmapKey){ return `${k.exchange}:${k.symbol}:${k.mode}`; }

  private get_or_create(key:HeatmapKey): ShaderHeatmapRenderer {
    const s=this.keyToString(key);
    let r=this.heatmaps.get(s);
    if (!r){
      r=new ShaderHeatmapRenderer();
      this.heatmaps.set(s, r);
    }
    return r;
  }

  apply_snapshot(pair:{exchange:string; symbol:string}, snapshot_pb:any, timeframe_seconds=0){
    const key:HeatmapKey={exchange: pair.exchange, symbol: pair.symbol, mode: snapshot_pb.mode};
    const reconstructor=this.get_or_create(key);
    if (timeframe_seconds>0 && timeframe_seconds*1000 !== reconstructor.get_column_interval_ms()) return;
    reconstructor.process_snapshot(snapshot_pb);
  }

  finalize_snapshot(pair:{exchange:string; symbol:string}, snapshot_pb:any){
    const mode=snapshot_pb.mode;
    let reconstructor=this.get_reconstructor(pair, mode);
    if (!reconstructor || !reconstructor.has_data()){
      this.apply_snapshot(pair, snapshot_pb);
      return;
    }
    const price_qty_map=new Map<number, number>();
    for (let i=0;i<snapshot_pb.prices.length;i++){
      const price=snapshot_pb.prices[i];
      const total_qty=snapshot_pb.bid_qty[i]+snapshot_pb.ask_qty[i];
      if (total_qty>=0.001) price_qty_map.set(price, total_qty);
    }
    reconstructor.finalize_column(snapshot_pb.timestamp_ms, price_qty_map);
  }

  get_reconstructor(pair:{exchange:string; symbol:string}, mode:string){
    const key:HeatmapKey={exchange: pair.exchange, symbol: pair.symbol, mode};
    return this.heatmaps.get(this.keyToString(key)) ?? null;
  }

  get_oldest_timestamp(pair:{exchange:string; symbol:string}, mode:string){
    const r=this.get_reconstructor(pair, mode);
    return r ? r.get_min_time() : 0;
  }
  get_newest_timestamp(pair:{exchange:string; symbol:string}, mode:string){
    const r=this.get_reconstructor(pair, mode);
    return r ? r.get_max_time() : 0;
  }
  has_data(pair:{exchange:string; symbol:string}, mode:string){
    const key:HeatmapKey={exchange: pair.exchange, symbol: pair.symbol, mode};
    return this.heatmaps.has(this.keyToString(key));
  }
  clear(pair:{exchange:string; symbol:string}, mode:string){
    const key:HeatmapKey={exchange: pair.exchange, symbol: pair.symbol, mode};
    const it=this.heatmaps.get(this.keyToString(key));
    if (it) it.clear();
  }
  clear_all(){ this.heatmaps.clear(); }
  mark_dirty(pair:{exchange:string; symbol:string}){
    for (const [k, reconstructor] of this.heatmaps){
      if (k.startsWith(`${pair.exchange}:${pair.symbol}:`)) reconstructor.mark_dirty();
    }
  }
  set_timeframe(pair:{exchange:string; symbol:string}, mode:string, timeframe_seconds:number){
    this.get_or_create({exchange: pair.exchange, symbol: pair.symbol, mode}).set_column_interval_ms(timeframe_seconds*1000);
  }
}

File: docs/edgedepth_original/src/core/heatmap_manager.cpp
#include "heatmap_manager.h"

ShaderHeatmapRenderer& HeatmapManager::get_or_create(const HeatmapKey& key) {
    auto it = heatmaps_.find(key);
    if (it == heatmaps_.end()) {
        auto renderer = std::make_unique<ShaderHeatmapRenderer>();
        it = heatmaps_.emplace(key, std::move(renderer)).first;
    }
    return *it->second;
}

// Called for HISTORICAL snapshots - builds the grid
void HeatmapManager::apply_snapshot(const Terminal::Pair& pair, const pb::HeatmapSnapshot& snapshot_pb, int64_t timeframe_seconds) {
    const HeatmapKey key{pair.exchange, pair.symbol, snapshot_pb.mode()};
    auto& reconstructor = get_or_create(key);
    if (timeframe_seconds > 0 && timeframe_seconds * 1000 != reconstructor.get_column_interval_ms()) return;
    // Use process_snapshot which builds/expands the grid
    // This is the method that initializes native_bucket_size_, bounds, etc.
    reconstructor.process_snapshot(snapshot_pb);
}

// Called for LIVE snapshots from server - finalizes a column (authoritative)
void HeatmapManager::finalize_snapshot(const Terminal::Pair& pair, const pb::HeatmapSnapshot& snapshot_pb) {
    const std::string mode = snapshot_pb.mode();
    auto* reconstructor = get_reconstructor(pair, mode);
    // No grid yet (live column arrived before any historical batch - e.g. a
    // pack replay whose backfill hasn't been requested/served, or a v1 pack
    // with no heatmap seed at all): BOOTSTRAP the grid from this column via
    // process_snapshot instead of dropping it. Before this, the book-depth
    // heatmap stayed permanently empty in pack mode because the historical
    // request was swallowed and every live column hit this early-return.
    if (!reconstructor || !reconstructor->has_data()) {
        apply_snapshot(pair, snapshot_pb);
        return;
    }
    // Convert protobuf to native map
    std::unordered_map<double, float> price_qty_map;
    price_qty_map.reserve(snapshot_pb.prices_size());

    for (int i = 0; i < snapshot_pb.prices_size(); ++i) {
        const double price = snapshot_pb.prices(i);
        const float total_qty = snapshot_pb.bid_qty(i) + snapshot_pb.ask_qty(i);
        if (total_qty >= 0.001f) {
            price_qty_map[price] = total_qty;
        }
    }

    // Finalize the column (authoritative server data)
    reconstructor->finalize_column(snapshot_pb.timestamp_ms(), price_qty_map);
}

ShaderHeatmapRenderer* HeatmapManager::get_reconstructor(const Terminal::Pair& pair, const std::string& mode) {
    const HeatmapKey key{pair.exchange, pair.symbol, mode};
    const auto it = heatmaps_.find(key);
    return (it != heatmaps_.end()) ? it->second.get() : nullptr;
}

int64_t HeatmapManager::get_oldest_timestamp(const Terminal::Pair& pair, const std::string& mode) {
    const auto* r = get_reconstructor(pair, mode);
    return r ? r->get_min_time() : 0;
}

int64_t HeatmapManager::get_newest_timestamp(const Terminal::Pair& pair, const std::string& mode) {
    const auto* r = get_reconstructor(pair, mode);
    return r ? r->get_max_time() : 0;
}

bool HeatmapManager::has_data(const Terminal::Pair& pair, const std::string& mode) const {
    const HeatmapKey key{pair.exchange, pair.symbol, mode};
    return heatmaps_.contains(key);
}

void HeatmapManager::clear(const Terminal::Pair& pair, const std::string& mode) {
    const HeatmapKey key{pair.exchange, pair.symbol, mode};
    auto it = heatmaps_.find(key);
    if (it != heatmaps_.end()) {
        it->second->clear();
    }
}

void HeatmapManager::clear_all() {
    heatmaps_.clear();
}

void HeatmapManager::mark_dirty(const Terminal::Pair& pair) {
    for (auto& [key, reconstructor] : heatmaps_) {
        if (key.exchange == pair.exchange && key.symbol == pair.symbol) {
            reconstructor->mark_dirty();
        }
    }
}
void HeatmapManager::set_timeframe(const Terminal::Pair& pair, const std::string& mode, int64_t timeframe_seconds) {
    get_or_create({pair.exchange, pair.symbol, mode}).set_column_interval_ms(timeframe_seconds * 1000);
}
